File: apps/ios/Pebbles/Features/Glyph/Views/GlyphBanner.swift
import SwiftUI

/// Centered identity banner: a glyph above a title + optional subtitle. Shared by
/// the profile header (`ProfileBanner` — glyph, display name, "Member since …")
/// and the glyph swap drawer (glyph, name, creator handle). Falls back to the
/// `.carve` placeholder when there are no strokes.
struct GlyphBanner: View {
    /// A byline renders its name half in the handwritten face (a name is
    /// always hand — issue #515); meta is a plain small-caps line.
    enum Subtitle {
        case meta(String)
        case byline(name: String)
    }

    let strokes: [GlyphStroke]?
    let title: String
    /// Token for the title. Profile uses `.largeTitleHand`; the glyph swap
    /// drawer keeps the serif `.title` for the glyph name.
    var titleFont: PebblesFont = .title
    let subtitle: Subtitle?

    var body: some View {
        VStack(spacing: Spacing.xxl) {
            glyph

            VStack(spacing: Spacing.xs) {
                Text(title)
                    .pebblesFont(titleFont)
                    .foregroundStyle(Color.system.foreground)
                subtitleView
            }
        }
        .frame(maxWidth: .infinity)
    }

    @ViewBuilder
    private var subtitleView: some View {
        switch subtitle {
        case .none:
            EmptyView()
        case .meta(let text):
            Text(text)
                .pebblesFont(.meta)
                .foregroundStyle(Color.system.secondary)
        case .byline(let name):
            HStack(alignment: .firstTextBaseline, spacing: Spacing.xs) {
                Text("BY")
                    .pebblesFont(.meta)
                    .foregroundStyle(Color.system.secondary)
                Text(name)
                    .pebblesFont(.bodyLeadHand)
                    .foregroundStyle(Color.system.secondary)
            }
        }
    }

    @ViewBuilder
    private var glyph: some View {
        if let strokes, !strokes.isEmpty {
            GlyphView(case: .profile, strokes: strokes, side: 96)
        } else {
            GlyphView(case: .carve, side: 96)
        }
    }
}

#Preview {
    GlyphBanner(
        strokes: [GlyphStroke(d: "M40,40 L160,160", width: 6)],
        title: "Creature",
        subtitle: .byline(name: "Galadriel")
    )
    .padding()
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/glyph/views/GlyphBanner.kt
package app.pbbls.android.features.glyph.views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import app.pbbls.android.R
import app.pbbls.android.features.glyph.models.GlyphStroke
import app.pbbls.android.theme.PebblesText
import app.pbbls.android.theme.PebblesTheme
import app.pbbls.android.theme.PebblesTypography

/**
 * Centered identity banner: a glyph above a title + optional subtitle — ports
 * iOS `Features/Glyph/Views/GlyphBanner.swift`. Shared by the profile header
 * (glyph, display name, "Member since …") and the glyph swap drawer (glyph,
 * name, creator byline). Falls back to the [GlyphViewCase.CARVE] placeholder
 * when there are no strokes.
 */
@Composable
fun GlyphBanner(
    title: String,
    modifier: Modifier = Modifier,
    strokes: List<GlyphStroke>? = null,
    viewBox: String = "0 0 200 200",
    // Profile uses largeTitleHand; the swap drawer keeps the serif-slot title.
    titleStyle: TextStyle = PebblesTypography.title,
    subtitle: GlyphBannerSubtitle? = null,
) {
    val system = PebblesTheme.colors.system
    Column(
        modifier = modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(PebblesTheme.spacing.xxl),
    ) {
        if (strokes.isNullOrEmpty()) {
            GlyphView(case = GlyphViewCase.CARVE, side = 96.dp)
        } else {
            GlyphView(case = GlyphViewCase.PROFILE, strokes = strokes, viewBox = viewBox, side = 96.dp)
        }
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(PebblesTheme.spacing.xs),
        ) {
            PebblesText(text = title, style = titleStyle, color = system.foreground)
            when (subtitle) {
                null -> Unit
                is GlyphBannerSubtitle.Meta ->
                    PebblesText(
                        text = subtitle.text,
                        style = PebblesTypography.meta,
                        color = system.secondary,
                    )
                is GlyphBannerSubtitle.Byline ->
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(PebblesTheme.spacing.xs),
                        verticalAlignment = Alignment.Bottom,
                    ) {
                        PebblesText(
                            text = stringResource(R.string.glyph_banner_by),
                            style = PebblesTypography.meta,
                            color = system.secondary,
                        )
                        // A name is always hand — issue #515.
                        PebblesText(
                            text = subtitle.name,
                            style = PebblesTypography.bodyLeadHand,
                            color = system.secondary,
                        )
                    }
            }
        }
    }
}

/** A byline renders its name in the handwritten face; meta is a plain small-caps line. */
sealed interface GlyphBannerSubtitle {
    data class Meta(
        val text: String,
    ) : GlyphBannerSubtitle

    data class Byline(
        val name: String,
    ) : GlyphBannerSubtitle
}
